agents/dqn.py

```diff
@@ -1,24 +1,132 @@
+from typing import Dict, Any
+
+import torch
+from torch import Tensor
+from torch.distributions import Categorical
+
 from agents import TorchAgent
+from ml.dqn_encoders import BaseQNetwork
 from ml.encoders import BaseEncoder
+from ml.utils import TensorWithMask
+from utils.bag_trajectory import BaseBagTrajectoryMemory
 
 
-# TODO[Vladimir Baikalov]: Implement
-class DQNAgent(TorchAgent, config_name='dqn'):
+def _with_random_research(
+        argmax_next_neighbor: Tensor,
+        neighbor: TensorWithMask,
+        prob: float
+):
+    mask = torch.bernoulli(torch.full(argmax_next_neighbor.shape, prob)).int()
```

Comment:
Btw, why do you want to use Bernoulli here? Afaik, it is possible to get the exact prop by applying `torch.softmax` to scores from your `q_network`.

Reply:
Prob is prob of random search, it doesn't concern q_values

Reply:
Ok, but afaik, we decided to implement solution with `torch.distributions.Categorical`. Correct me if I am wrong

Reply:
done)